full-service/src/service/password_manager.rs

```diff
@@ -0,0 +1,440 @@
+use blake2::{Blake2b, Digest};
+use diesel::Connection;
+use displaydoc::Display;
+
+const SALT_DOMAIN_TAG: &str = "full-service-salt";
+
+#[derive(Display, Debug)]
+pub enum PasswordServiceError {
+    /// Error interacting with the database: {0}
+    Database(WalletDbError),
+
+    /// Error decoding from hex: {0}
+    HexDecode(hex::FromHexError),
+
+    /// Cannot set password on encrypted database. Must change_password.
+    DatabaseEncrypted,
+
+    /// Cannot perform this action without a set password or while database is locked. Please set_password or unlock first.
+    DatabaseLocked,
+
+    /// Must provide either password or password hash, not both.
+    CannotDisambiguatePassword,
+
+    /// Argon2 Error: {0}
+    Argon2(argon2::Error),
+
+    /// Diesel Error: {0}
+    Diesel(diesel::result::Error),
+}
+
+impl From<WalletDbError> for PasswordServiceError {
+    fn from(src: WalletDbError) -> Self {
+        Self::Database(src)
+    }
+}
+
+impl From<hex::FromHexError> for PasswordServiceError {
+    fn from(src: hex::FromHexError) -> Self {
+        Self::HexDecode(src)
+    }
+}
+
+impl From<argon2::Error> for PasswordServiceError {
+    fn from(src: argon2::Error) -> Self {
+        Self::Argon2(src)
+    }
+}
+
+impl From<diesel::result::Error> for PasswordServiceError {
+    fn from(src: diesel::result::Error) -> Self {
+        Self::Diesel(src)
+    }
+}
+
+pub trait PasswordService {
+    // Helper method to expand password to password hash using argon2.
+    fn get_password_hash(
+        &self,
+        password: Option<String>,
+        password_hash: Option<String>,
+    ) -> Result<Vec<u8>, PasswordServiceError>;
+
+    /// The initial call to set the password for the DB.
+    fn set_password(
+        &self,
+        password: Option<String>,
+        password_hash: Option<String>,
+    ) -> Result<bool, PasswordServiceError>;
+
+    /// Unlock the DB.
+    fn unlock(
+        &self,
+        password: Option<String>,
+        password_hash: Option<String>,
+    ) -> Result<bool, PasswordServiceError>;
+
+    /// Change the password for the DB.
+    fn change_password(
+        &self,
+        old_password: Option<String>,
+        old_password_hash: Option<String>,
+        new_password: Option<String>,
+        new_password_hash: Option<String>,
+    ) -> Result<bool, PasswordServiceError>;
+
+    /// Whether the database is locked.
+    ///
+    /// Returns:
+    /// * Some(true) if database is locked
+    /// * Some(false) if database is unlocked
+    /// * None if database has not yet had a password set up.
+    fn is_locked(&self) -> Result<Option<bool>, PasswordServiceError>;
+
+    /// Utility method to call on service methods which require the database to be unlocked
+    fn verify_unlocked(&self) -> Result<(), PasswordServiceError>;
+}
+
+impl<T, FPR> PasswordService for WalletService<T, FPR>
+where
+    T: BlockchainConnection + UserTxConnection + 'static,
+    FPR: FogPubkeyResolver + Send + Sync + 'static,
```

Comment:
Is this needed?

Reply:
I believe so - without it, I get "Wrong number of type arguments"